pkg/volume/csimigration/plugin_manager.go

```diff
@@ -0,0 +1,154 @@
+	case csilibplugins.GCEPDInTreePluginName:
+		return utilfeature.DefaultFeatureGate.Enabled(features.CSIMigrationGCE)
+	case csilibplugins.AzureFileInTreePluginName:
+		return utilfeature.DefaultFeatureGate.Enabled(features.CSIMigrationAzureFile)
+	case csilibplugins.AzureDiskInTreePluginName:
+		return utilfeature.DefaultFeatureGate.Enabled(features.CSIMigrationAzureDisk)
+	case csilibplugins.CinderInTreePluginName:
+		return utilfeature.DefaultFeatureGate.Enabled(features.CSIMigrationOpenStack)
+	default:
+		return false
+	}
+}
+
+// IsMigratable indicates whether CSI migration has been enabled for a volume
+// plugin that the spec refers to
+func (pm PluginManager) IsMigratable(spec *volume.Spec) (bool, error) {
+	if spec == nil {
+		return false, fmt.Errorf("could not find if plugin is migratable because volume spec is nil")
+	}
+
+	pluginName, _ := pm.GetInTreePluginNameFromSpec(spec.PersistentVolume, spec.Volume)
+	if pluginName == "" {
+		return false, nil
+	}
+	// found an in-tree plugin that supports the spec
+	return pm.IsMigrationEnabledForPlugin(pluginName), nil
+}
+
+// InTreeToCSITranslator performs translation of Volume sources for PV and Volume objects
+// from references to in-tree plugins to migrated CSI plugins
+type InTreeToCSITranslator interface {
+	TranslateInTreePVToCSI(pv *v1.PersistentVolume) (*v1.PersistentVolume, error)
+	TranslateInTreeInlineVolumeToCSI(volume *v1.Volume) (*v1.PersistentVolume, error)
+}
+
+// TranslateInTreeSpecToCSI translates a volume spec (either PV or inline volume)
+// supported by an in-tree plugin to CSI
+func TranslateInTreeSpecToCSI(spec *volume.Spec, translator InTreeToCSITranslator) (*volume.Spec, error) {
+	var csiPV *v1.PersistentVolume
+	var err error
+	inlineVolume := false
+	if spec.PersistentVolume != nil {
+		csiPV, err = translator.TranslateInTreePVToCSI(spec.PersistentVolume)
+	} else if spec.Volume != nil {
+		csiPV, err = translator.TranslateInTreeInlineVolumeToCSI(spec.Volume)
+		inlineVolume = true
+	} else {
+		err = errors.New("not a valid volume spec")
+	}
+	if err != nil {
+		return nil, fmt.Errorf("failed to translate in-tree pv to CSI: %v", err)
+	}
+	return &volume.Spec{
+		PersistentVolume:                csiPV,
+		ReadOnly:                        spec.ReadOnly,
+		InlineVolumeSpecForCSIMigration: inlineVolume,
+	}, nil
+}
+
+// CheckMigrationFeatureFlags checks the configuration of feature flags related
+// to CSI Migration is valid
+func CheckMigrationFeatureFlags(f featuregate.FeatureGate, pluginMigration, pluginMigrationComplete featuregate.Feature) error {
+	if f.Enabled(pluginMigration) && !f.Enabled(features.CSIMigration) {
+		return fmt.Errorf("enabling %q requires CSIMigration to be enabled", pluginMigration)
+	}
+	if f.Enabled(pluginMigrationComplete) && !f.Enabled(pluginMigration) {
+		return fmt.Errorf("enabling %q requires %q to be enabled", pluginMigrationComplete, pluginMigration)
+	}
+	return nil
+}
```

Comment:
granted the "turn on all alpha features" CI job is an unusual case, but I'm reconsidering making this a failure given the gymnastics https://github.com/kubernetes/test-infra/pull/15286/files has to go through. Maybe just warning in these cases would be sufficient.